src/main/services/apiService.js

```diff
@@ -0,0 +1,164 @@
+'use strict';
+
+function ApiService($http, apiEndpoint, apiEndpointRedirect, $q) {
+  this.http = $http;
+  this.apiEndpoint = apiEndpoint;
+  this.apiEndpointRedirect = apiEndpointRedirect;
+  this.q = $q;
+}
+
+ApiService.prototype.request = function (param) {
+  return this.http.get(this.apiEndpoint + param).then(function (response) {
+    return response.data;
+  });
+};
+
+ApiService.prototype.getProjects = function(){
+  return this.request('projects.json').then(function (data) {
+    return data.projects;
+  });
+};
+
+ApiService.prototype.getVersions = function (project) {
+  return this.request('projects/' + project + '/versions.json').then(function (data) {
+    return data.versions;
+  });
+};
+
+ApiService.prototype.getIssues = function (config) {
+  var params = this.generateGeneralIssuesParameters(config);
+  var limit = config.limit ? config.limit : Number.MAX_SAFE_INTEGER;
+  return this.getIssuesWithParamsAndLimit(params, limit);
+};
+
+ApiService.prototype.getIssuesForChart = function(config) {
+  var allIssues = [];
+  var limit = config.limit ? config.limit : Number.MAX_SAFE_INTEGER;
+  var params1 = this.generateParametersForIssuesOpenOnEnd(config);
+  var params2 = this.generateParametersForIssuesClosedBetweenStartAndEnd(config);
+  var params3 = this.generateParametersForIssuesOpen(config);
+  return this.q.all([this.getIssuesWithParamsAndLimit(params1, limit), this.getIssuesWithParamsAndLimit(params2, limit),
+    this.getIssuesWithParamsAndLimit(params3, limit)]).then(function (responses) {
+    angular.forEach(responses, function (issues) {
+      angular.forEach(issues, function (issue) {
+        allIssues.push(issue);
+      });
+    });
+    return allIssues;
+  });
+};
+
+ApiService.prototype.getIssuesWithParamsAndLimit = function (params, limit) {
+  var allIssues = [];
+  var q = this.q;
+  return this.collectPageIssues(params, 0).then(function (issues) {
+    angular.forEach(issues.issues, function (issue) {
+      allIssues.push(issue);
+    });
+    var requests = [];
+    for (var i = 100; i < issues.total_count && i < limit; i = i + 100) {
+      requests.push(this.collectPageIssues(params, i));
```

Comment:
I think the call "this.collectPageIssues" throws an error because it is used in an anonymous function.

Reply:
Yeah, thanks. Of course, it does not know the prototype instance.